<a href="https://colab.research.google.com/github/lima-breno/natural_language_processing/blob/main/Lista2_NLP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#LISTA DE EXERCÍCIOS 02


## Importando bibliotecas

In [ ]:
import nltk
from nltk.corpus import machado
nltk.download('machado')
nltk.download('punkt_tab')
nltk.download('punkt')
nltk.download('stopwords')

[nltk_data] Downloading package machado to /root/nltk_data...
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

## Definindo base de dados

In [ ]:
contos = []
for k in range(1,9):
  contos.extend(machado.sents(f'romance/marm0{k}.txt'))

In [ ]:
from string import punctuation
import re
from nltk.corpus import stopwords

stopwords = stopwords.words('portuguese')

def pre_processamento(sentenca):
  sentenca = [token for token in sentenca if token not in punctuation]
  sentenca = [token.lower() for token in sentenca]
  sentenca = [re.sub(r'\d', '', token) for token in sentenca]
  sentenca = [token for token in sentenca if token not in stopwords]
  return sentenca

machado_memorias = [pre_processamento(sentenca) for sentenca in contos]

##Exercícios

###Exercício 01
Construa três modelos de embedding variando apenas o número de epochs nos valores: 5, 10 e 20. Use os seguintes parâmetros fixos para todos os modelos:

window = 3

sg = 0

min_count = 3

vector_size = 100

In [ ]:
from gensim.models import Word2Vec

# Lista de valores de epochs que serão testados
valores_de_epochs = [5, 10, 20]

# Parâmetros fixos
params_fixos = {
    "vector_size": 100,
    "window": 3,
    "sg": 0,
    "min_count": 3
}

# Dicionário para armazenar os modelos
modelos = {}

# Treinamento dos modelos
for ep in valores_de_epochs:
    print(f"Treinando modelo com {ep} epochs...")
    modelo = Word2Vec(
        sentences=machado_memorias,
        epochs=ep,
        **params_fixos
    )
    modelos[f"modelo_{ep}ep"] = modelo
    print(f"Modelo com {ep} epochs treinado com sucesso!\n")

Treinando modelo com 5 epochs...
Modelo com 5 epochs treinado com sucesso!

Treinando modelo com 10 epochs...
Modelo com 10 epochs treinado com sucesso!

Treinando modelo com 20 epochs...
Modelo com 20 epochs treinado com sucesso!



### Exercício 02

Para cada modelo, obtenha as 5 palavras mais similares às seguintes palavras:


*  amor
*  paixão
*  memória
*  relógio


In [ ]:
# Lista de palavras que queremos analisar
palavras_alvo = ["amor", "paixão", "memória", "relógio"]

# Para cada modelo treinado (com 5, 10 e 20 epochs)
for nome_modelo, modelo in modelos.items():
    print(f"\n Resultados para o {nome_modelo}:\n")

    # Para cada palavra da lista
    for palavra in palavras_alvo:
        print(f"Palavras parecidas com '{palavra}':")

        # Verifica se a palavra existe no vocabulário do modelo
        if palavra in modelo.wv:
            similares = modelo.wv.most_similar(palavra, topn=5)
            for sim_palavra, score in similares:
                print(f"  {sim_palavra}  →  similaridade: {score:.2f}")
        else:
            print(f"  (A palavra '{palavra}' não foi encontrada no vocabulário.)")

        print()  # Linha em branco para separar os resultados


 Resultados para o modelo_5ep:

Palavras parecidas com 'amor':
  verdade  →  similaridade: 1.00
  poderia  →  similaridade: 1.00
  homem  →  similaridade: 1.00
  tudo  →  similaridade: 1.00
  agora  →  similaridade: 1.00

Palavras parecidas com 'paixão':
  vista  →  similaridade: 1.00
  leitor  →  similaridade: 1.00
  aí  →  similaridade: 1.00
  confiança  →  similaridade: 1.00
  criança  →  similaridade: 1.00

Palavras parecidas com 'memória':
  bela  →  similaridade: 1.00
  acaso  →  similaridade: 1.00
  flor  →  similaridade: 1.00
  alegria  →  similaridade: 1.00
  passado  →  similaridade: 1.00

Palavras parecidas com 'relógio':
  primeiras  →  similaridade: 1.00
  trabalho  →  similaridade: 1.00
  sol  →  similaridade: 1.00
  umas  →  similaridade: 1.00
  caminho  →  similaridade: 1.00


 Resultados para o modelo_10ep:

Palavras parecidas com 'amor':
  sentimento  →  similaridade: 0.95
  natureza  →  similaridade: 0.95
  tal  →  similaridade: 0.94
  porque  →  similaridade: 0.94


### Exercício 03

### Quais diferenças você observou nas palavras mais similares conforme o número de epochs aumentou?**

À medida que o número de epochs aumenta, nota-se uma mudança significativa tanto nos valores de similaridade quanto na qualidade das palavras mais próximas:

Com 5 epochs:

Todas as palavras similares apresentam similaridade máxima (1.00), o que é atípico e sugere que o modelo ainda não capturou relações semânticas reais.

As palavras retornadas, em muitos casos, não têm relação semântica clara com as palavras-alvo (ex: "amor" = "verdade", "poderia", "homem", "tudo", "agora").

Com 10 epochs:

Os valores de similaridade diminuem levemente (em torno de 0.94 a 0.99), indicando que o modelo começa a diferenciar melhor as relações.

As palavras similares tornam-se mais relevantes semanticamente (ex: "amor" = "sentimento", "natureza"; "paixão" = "afeição", "confiança").

Com 20 epochs:

A similaridade diminui ainda mais (0.71 a 0.90), sugerindo maior refinamento dos vetores.

As palavras próximas são, em sua maioria, semanticamente relacionadas às palavras-alvo (ex: "amor" = "sentimento", "pura", "amava"; "paixão" = "afeição", "ambição", "confiança").

###Algum dos modelos produziu resultados mais coerentes ou interessantes? Justifique.**

O modelo com 20 épocas apresenta resultados ainda mais refinados, com associações que demonstram compreensão semântica mais profunda (ex: "amor" = "amava", "admiração", "vaidade"; "memória" = "responsabilidade", "mocidade", "mistura").
As palavras similares são mais específicas e contextualizadas, o que pode ser considerado mais interessante para análises literárias.

###Com base nas análises, qual valor de epoch pareceu mais apropriado para este conjunto de dados?**

O valor de 20 epochs parece ser o mais apropriado para este conjunto de dados, pois produz resultados mais coerentes, específicos e semanticamente interessantes, refletindo melhor as nuances do texto literário analisado. Isso é especialmente desejável em tarefas de análise textual onde a riqueza semântica é fundamental.